crates/sui-framework/sources/display.move

```diff
@@ -0,0 +1,184 @@
+/// substitution and filling-in the pieces using the data from the object T.
+///
+/// More entry functions might be added in the future depending on the use cases.
+module sui::display {
+    use sui::publisher::{is_package, Publisher};
+    use sui::tx_context::TxContext;
+    use std::string::{String, utf8};
+    use sui::vec_map::{Self, VecMap};
+    use sui::object::{Self, ID, UID};
+    use sui::transfer;
+    use sui::event;
+    use std::vector;
+
+    /// For when T does not belong to package in PublisherCap.
+    const ENotOwner: u64 = 0;
+
+    /// For when vectors passed into one of the multiple insert functions
+    /// don't match in their lengths.
+    const EVecLengthMismatch: u64 = 1;
+
+    /// The Display object. Defines the way an object should be
+    /// displayed. Display object can only be created and modified with
+    /// a PublisherCap, making sure that the rules are set by the owner
+    /// of the type.
+    ///
+    /// Each of the display properties should support patterns outside
+    /// of the system, making it simpler to customize Display based
+    /// on the property values of an Object.
+    /// ```
+    /// // Example of a display object
+    /// Display<0x...::capy::Capy> {
+    ///  fields:
+    ///    <name, "Capy {{ genes }}">
+    ///    <link, "https://capy.art/capy/{{ id }}">
+    ///    <image, "https://api.capy.art/capy/{{ id }}/svg">
+    ///    <description, "Lovely Capy, one of many">
+    /// }
+    /// ```
+    ///
+    /// Uses only String type due to external-facing nature of the object,
+    /// the property names have a priority over their types.
+    struct Display<phantom T: key> has key {
+        id: UID,
+        /// Contains fields for display. Currently supported
+        /// fields are: name, link, image and description.
+        fields: VecMap<String, String>,
+    }
+
+    /// Event: emitted when a new Display object has been created for type T.
+    /// Type signature of the event corresponds to the type while id serves for
+    /// the discovery.
+    struct DisplayCreated<phantom T: key> has copy, drop {
+        id: ID
+    }
+
+    // ============ Initializer Functions ===========
+
+    /// Since the only way to own a Display is before it has been published,
+    /// we don't need to perform an authorization check.
+    ///
+    /// Also, the only place it can be used is the function where the Display
+    /// object was created; hence values and names are likely to be hardcoded and
+    /// vector<u8> is the best type for that purpose.
+    public fun set_owned<T: key>(d: Display<T>, name: vector<u8>, value: vector<u8>): Display<T> {
+        set_internal(&mut d, utf8(name), utf8(value));
+        d
+    }
+
+    /// Create an empty Display object. It can either be shared empty of filled
+    /// with data right away via cheaper `set_owned` method.
+    public fun empty<T: key>(pub: &Publisher, ctx: &mut TxContext): Display<T> {
```

Comment:
Should this specify that it should be called at most once for a given type `T` (and explain what happens if that assumption is violated)?

Reply:
That's a good one! I have an idea of creating a centralized registry of displays for the sake of discoverability and also uniqueness. We might handle versions there as well. 

Another benefit we can get from the centralization is the ability to destroy older versions of the display and get a storage rebate.

Reply:
Can you update the comment to explain the intention?